flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
 
         self.assertRaises(UnattachedVolume,
                           self.cache.get_device_path, attached_id1)
+
+
+@attributes(["_proxy_object", "_callback"], apply_with_init=False)
+class CallbackProxy(object):
+    """
+    A proxy to another object that calls a callback before and after each call
+    to the underlying object. This enables tests where you want to inject code
+    before or after any calls to an underlying API.
+
+    Note that this proxy assumes that the underlying object only has public
+    methods and not public attributes.
+
+    :ivar _proxy_object: The object that is being proxied to.
+    :ivar _callback: The callback to be called before and after every call to a
+        method of _proxy_object.
+    """
+
+    def __init__(self, proxy_object, callback):
+        self._proxy_object = proxy_object
+        self._callback = callback
+
+    def __getattr__(self, name):
+        """
+        The implementation of the proxy. Gets the attribute on the underlying
+        object and returns a wrapped function that calls the callback before
+        and after its execution.
+
+        :param name: The name of the method on the proxy object that should be
+            called.
+        """
+        method = getattr(self._proxy_object, name)
+
+        def proxied_function(*args, **kwargs):
+            self._callback('Before %s' % name)
+            result = method(*args, **kwargs)
+            self._callback('After %s' % name)
+            return result
+
+        return proxied_function
+
+
+class EndToEndBlockdeviceDeployerTests(AsyncTestCase, ScenarioMixin):
+    """
+    Tests that only interact with the BlockDeviceDeployer by changing the
+    desired configuration of the cluster, and by interfacing with the datasets
+    that are created locally.
+    """
+
+    def _change_node_state(self, deployer, desired_configuration,
+                           cluster_state_callback):
+        """
+        Change the local state to match the given desired configuration. Will
+        run the covergence loop until it determines that there is nothing left
+        to do.
+
+        :param IDeployer deployer: Deployer to discover local state and
+            calculate changes.
+        :param Deployment desired_configuration: The intended configuration of
+            all nodes.
+        :param cluster_state_callback: A callable that will be called with a
+            single argument, the cluster state, after every discover_state
+            cycle.
+
+        :return: ``Deferred`` that fires when the necessary changes are done.
+        """
+        # XXX: Note to the reviewer - I wish I could use something closer to
+        # the actaul production code here, but I didn't see how to do that. Do
+        # you? The one exception is that I want there to be no state persisted
+        # from one run of converge to the next.
```

Comment:
> If you mean explicit states of datasets, I'm sort of trying to avoid having to depend on those.

I do mean that, and I don't think it makes sense to try avoid depending on them. I think it much easier to reason about them that it is to reason about cluster configurations (and in particular, easier to see that you have full coverage of possible states and state transitions.